File: IvanShurenkov/Second Term/Block1/Task1/src/main/kotlin/filters/GrayFilter.kt
package filters

import BMPFile

class GrayFilter : Filter() {
    private val redCoefficient: Double = 0.2126
    private val greenCoefficient: Double = 0.7152
    private val blueCoefficient: Double = 0.0722

    fun applyFilter(bmpFile: BMPFile) {
        for (i in 0 until bmpFile.height) {
            for (j in 0 until (bmpFile.row - 2) step (3 + (bmpFile.bitCnt == 32).toInt())) {
                val gray: Int = ((bmpFile.image[i][j] * redCoefficient +
                        bmpFile.image[i][j + 1] * greenCoefficient +
                        bmpFile.image[i][j + 2] * blueCoefficient).toInt() and 0xff)
                bmpFile.image[i][j] = gray.toShort()
                bmpFile.image[i][j + 1] = gray.toShort()
                bmpFile.image[i][j + 2] = gray.toShort()
            }
        }
    }
}

File: Oleg-Tsvetkov/Term 2/Block 1/BMPFilter/BMPFilter/Filters/GrayFilter.cs
namespace BMPFilter.Filters
{
    public static class GrayFilter
    {
        private const double RedCoefficient = 0.3;
        private const double GreenCoefficient = 0.59;
        private const double BlueCoefficient = 0.11;

        public static void ApplyFilter(BitMapFile image)
        {
            for (int i = 0; i < image.Height; ++i)
            {
                for (int j = 0; j < image.RowByteSize; j += image.Channels)
                {
                    // Colors are stored in BGR order
                    byte color = (byte)(BlueCoefficient * image.ImageBytes[i][j] +
                                         GreenCoefficient * image.ImageBytes[i][j + 1] +
                                         RedCoefficient * image.ImageBytes[i][j + 2]);
                    image.ImageBytes[i][j] = color;
                    image.ImageBytes[i][j + 1] = color;
                    image.ImageBytes[i][j + 2] = color;
                }
            }
        }
    }
}